File: packages/guards/src/instruction-boundary.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface InstructionBoundaryOptions {
  action: 'block' | 'warn';
  /** System instructions that should not leak */
  systemInstructions?: string[];
  /** Detect attempts to extract system prompt */
  detectExtraction?: boolean;
}

const EXTRACTION_PATTERNS = [
  /\b(?:what|show|reveal|display|print|output|repeat|echo)\s+(?:your|the)\s+(?:system|initial|original)\s+(?:prompt|instruction|message)/i,
  /\b(?:tell|give|share)\s+me\s+(?:your|the)\s+(?:system|initial|original)\s+(?:prompt|instruction)/i,
  /\bsystem\s*prompt\s*[:=]/i,
  /\b(?:above|previous|prior|initial)\s+(?:instructions?|prompt|context)\b/i,
  /\brepeat\s+(?:everything|all|the\s+text)\s+(?:above|before|from\s+the\s+beginning)/i,
  /\bwhat\s+(?:were|are)\s+you\s+(?:told|instructed|programmed)/i,
];

export function instructionBoundary(options: InstructionBoundaryOptions): Guard {
  const detectExtraction = options.detectExtraction ?? true;
  const systemInstructions = (options.systemInstructions ?? []).map((s) => s.toLowerCase());

  return {
    name: 'instruction-boundary',
    version: '0.1.0',
    description: 'Prevents system instruction leakage and extraction attempts',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const violations: string[] = [];

      if (detectExtraction) {
        for (const pat of EXTRACTION_PATTERNS) {
          const m = text.match(pat);
          if (m) {
            violations.push(`Extraction attempt: ${m[0]}`);
          }
        }
      }

      if (systemInstructions.length > 0) {
        const lower = text.toLowerCase();
        for (const instr of systemInstructions) {
          if (lower.includes(instr)) {
            violations.push('System instruction content detected in text');
            break;
          }
        }
      }

      const triggered = violations.length > 0;
      return {
        guardName: 'instruction-boundary',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { violations } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/instruction_boundary.py
"""Prevents system instruction leakage and extraction."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_EXTRACTION_PATTERNS = [
    re.compile(r"\b(?:what|show|reveal|display|print|output|repeat|echo)\s+(?:your|the)\s+(?:system|initial|original)\s+(?:prompt|instruction|message)", re.I),
    re.compile(r"\b(?:tell|give|share)\s+me\s+(?:your|the)\s+(?:system|initial)\s+(?:prompt|instruction)", re.I),
    re.compile(r"\bsystem\s*prompt\s*[:=]", re.I),
    re.compile(r"\brepeat\s+(?:everything|all|the\s+text)\s+(?:above|before|from\s+the\s+beginning)", re.I),
    re.compile(r"\bwhat\s+(?:were|are)\s+you\s+(?:told|instructed|programmed)", re.I),
]


class _InstructionBoundary:
    def __init__(self, *, action: str = "block", system_instructions: Optional[List[str]] = None, detect_extraction: bool = True) -> None:
        self.name = "instruction-boundary"
        self.action = action
        self._instructions = [s.lower() for s in (system_instructions or [])]
        self._detect = detect_extraction

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        violations: List[str] = []
        if self._detect:
            for p in _EXTRACTION_PATTERNS:
                m = p.search(text)
                if m:
                    violations.append(f"Extraction attempt: {m.group()[:60]}")
        if self._instructions:
            lower = text.lower()
            for instr in self._instructions:
                if instr in lower:
                    violations.append("System instruction content in text")
                    break
        triggered = len(violations) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(guard_name="instruction-boundary", passed=not triggered, action=self.action if triggered else "allow", message=violations[0] if triggered else None, latency_ms=round(elapsed, 2), details={"violations": violations} if triggered else None)


def instruction_boundary(*, action: str = "block", system_instructions: Optional[List[str]] = None, detect_extraction: bool = True) -> _InstructionBoundary:
    return _InstructionBoundary(action=action, system_instructions=system_instructions, detect_extraction=detect_extraction)
